Contributing towards issue #419 "Requirements for list of mentors/students"

chandan00761 opened the pull request:
# Commits

+added GSoCYearFilter which filters users by year with current year as default value.

! changed get_queryset of HiddenUserProfileAdmin such that:
  * when request.user is superadmin all users are returned
  * when request.user is Suborg Admin users of that suborg are returned.
  * else empty query is returned.

# Description

[changes described here](https://github.com/python-gsoc/python-blogs/issues/419#issuecomment-982673691)

Fixes #419 

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update.

# How Has This Been Tested?

**To make this feature work do the following steps:**
* Create a new group suborg_admin and provide permission "gsoc | user profile | Can view user profile".
* Add all the suborg admins to this group.

**Testing**
* Log in as any of the suborg admin and go to the admin screen.
* Pick **User Profiles** under **GSoC**.

- [x] Manual Testing

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have not added a commit to any .db files as part of my pull request
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

chandan00761 committed:
+ added GSoCYearFilter which filters users by year with current year as default value.
! changed get_queryset of HiddenUserProfileAdmin such that:
  * when request.user is superadmin all users are returned
  * when request.user is Suborg Admin users of that suborg are returned.
  * else empty query is returned.

botanicvelious commented on gsoc/admin.py:
```diff
@@ -1,4 +1,5 @@
 from .models import *
+from .filters import GSoCYearFilter
```
dont need to make a whole new file and class for this

chandan00761 replied:
Fixed

botanicvelious commented on gsoc/admin.py:
```diff
@@ -396,13 +397,23 @@ def email(self, obj):
+        user = request.user.userprofile_set.get()
+        if request.user.is_superuser:
+            return UserProfile.all_objects.all()
+        if user.role == 1:
```
so a user can have more then one role, so you will need to check all roles for the current year

chandan00761 replied:
if the user is superuser then we return all the UserProfiles sorted by latest year.
```
UserProfile.all_objects.all().order_by("gsoc_year")
```
else we check if any of the UserProfiles has role set as suborg admin
```
user_profiles = user.userprofile_set.filter(role=1).all()
```

botanicvelious commented on gsoc/filters.py:
```diff
@@ -0,0 +1,40 @@
+from django.contrib.admin import SimpleListFilter
+
+
+class GSoCYearFilter(SimpleListFilter):
```
This entire class isnt really necessary, no need to make a new file and class for one function, just stick this code in admin.py also doing it this way is like 2x the amount of code

chandan00761 replied:
Fixed

botanicvelious commented on gsoc/admin.py:
```diff
@@ -396,13 +397,23 @@ def email(self, obj):
+        if request.user.is_superuser:
+            return UserProfile.all_objects.all()
+        if user.role == 1:
+            return UserProfile.objects.filter(suborg_full_name=user.suborg_full_name)
```
you could set this filter to be the current year, as well as the suborg in question at the same time, doing so would make the filters.py redundant

chandan00761 replied:
For suborg admins only current year mentors are shown for psf admins all the users are shown sorted by year.

botanicvelious commented on gsoc/admin.py:
```diff
@@ -353,10 +354,10 @@ class HiddenUserProfileAdmin(admin.ModelAdmin):
         "gsoc_invited",
     )
-    list_filter = ("hidden", "reminder_disabled", "role", "gsoc_invited")
+    list_filter = ("hidden", "reminder_disabled", "role", "gsoc_invited", GSoCYearFilter)
```
As I mentioned, why have a new class? just display the data that is already there.

chandan00761 replied:
removed filter

botanicvelious commented on gsoc/admin.py:
```diff
@@ -396,13 +397,23 @@ def email(self, obj):
+        """
+        user = request.user.userprofile_set.get()
+        if request.user.is_superuser:
+            return UserProfile.all_objects.all()
```
this should be limited to the current year

botanicvelious reviewed (changes requested):
Let me know if you have any questions on my comments, for something like this its usually best to not make extra classes etc that will only be used once, especially when it can be done in the same filter query you are already using

chandan00761 commented:
Thank you for the feedback. I implement the feature in a more "beginner friendly" way hence more lines of code. I will make all the changes suggested and ping you back.

Changelist:

- [x] remove filters.py and GSoCYearFilter
- [x] add inline year filter.
- [x] check roles of all UserProfiles
- [x] All users of only current year should be shown to admin by default.

chandan00761 committed:
- removed GSoCYearFilter and filters.py
- removed gsoc_year for list_filter in HiddenuserProfileAdmin
! fixed issue where all UserProfiles associated with user where not checked.

pep8speaks commented:
Hello @chandan00761! Thanks for updating this PR. 

* In the file [`gsoc/admin.py`](https://github.com/python-gsoc/python-blogs/blob/707f5498a888626defd0cee94f5f8141b9b3234e/gsoc/admin.py):

> [Line 423:101](https://github.com/python-gsoc/python-blogs/blob/707f5498a888626defd0cee94f5f8141b9b3234e/gsoc/admin.py#L423): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (114 > 100 characters)





##### Comment last updated at 2021-12-13 10:58:53 UTC

chandan00761 committed:
! fixed pep8 errors

chandan00761 commented:
I have done the changes requested. Please review if it gets the work done.

botanicvelious commented:
Took a quick look, the overall logic looks right now, just 2 questions

1) why is this a change to get_queryset in the HiddenUserProfileAdmin class?
2) how will users be able to see this list? should probably add a menu item for suborg admins for it

chandan00761 commented:
1. Since only a few conditions needed checking to return the filtered Queryset, I have done it inside the get_queryset only.
2. Suborg admin can access the list from UserProfiles menu item only. Only thing they should have is, access to the Admin dashboard and  "gsoc | userprofiles | can view user profile" permission.

botanicvelious commented:
My concern with editing the other class is that class may be used elsewhere, I dont see it being used anywhere however so not sure why it existed before but im likely missing it.

For the permission that is required then we need code added for adding the permission automatically dont we? were you able to test the code?

chandan00761 committed:
+ added receiver to change user permission whenever any user profile associated with it is changed.

chandan00761 committed:
+ used has_perm to check for permission

chandan00761 commented:
@botanicvelious
